src/main.cpp

```diff
@@ -1074,8 +1111,35 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
         }
     }
 
+    // Reject Overwintered transactions which have expired
+    if (tx.fOverwintered &&
+        tx.nVersion == 3 &&
+        tx.nFormatBranchId != 0 &&
+        tx.nExpiryHeight != std::numeric_limits<uint32_t>::max())
+    {
+        // FIXME: Narrowing from using cast because types mismatched with expiry height as uint32, height as int
+        if ((int)tx.nExpiryHeight <= chainActive.Height()) {
+            return state.DoS(100, error("AcceptToMemoryPool: tx expired at block height: %u", tx.nExpiryHeight),
+                             REJECT_INVALID, "tx-expired");
+        }
+    }
+
+    // FIXME: Replace 0xDEADBEEF with NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId;
+    if (tx.fOverwintered &&
+        (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION)) {
+        return state.DoS(100, error("CheckTransaction(): unknown overwintered version"),
+            REJECT_INVALID, "unknown-overwintered-version");
+    }
+
+    uint32_t consensusBranchId = 0;
+    // FIXME: get the branch id given current block height... for now set to 0xDEADBEEF
+    if (tx.fOverwintered && tx.nVersion >= 3) {
```

Comment:
We must not set `consensusBranchId` based on transaction information, as that creates a circular consensus rule dependency. Use the branch ID that will be associated with the next block that the transaction could be mined in:

```
auto consensusBranchId = CurrentEpochBranchId(chainActive.Height() + 1, Params().GetConsensus());
```

Reply:
Removed conditional.  Added check to enforce `fOverwintered` must be true if Overwinter is active.